lldb/source/Target/SwiftLanguageRuntimeDynamicTypeResolution.cpp

```diff
@@ -34,21 +34,22 @@ using namespace lldb;
 using namespace lldb_private;
 
 namespace lldb_private {
-swift::Type GetSwiftType(void *opaque_ptr) {
-  return reinterpret_cast<swift::TypeBase *>(opaque_ptr);
-}
-
-swift::CanType GetCanonicalSwiftType(void *opaque_ptr) {
-  return reinterpret_cast<swift::TypeBase *>(opaque_ptr)->getCanonicalType();
-}
-
-swift::CanType GetCanonicalSwiftType(const CompilerType &type) {
-  return GetCanonicalSwiftType(
-      reinterpret_cast<void *>(type.GetOpaqueQualType()));
+swift::Type GetSwiftType(CompilerType type) {
+  auto *ts = type.GetTypeSystem();
+  if (auto *tr = llvm::dyn_cast_or_null<TypeSystemSwiftTypeRef>(ts))
+    return tr->GetSwiftType(type);
+  if (auto *ast = llvm::dyn_cast_or_null<SwiftASTContext>(ts))
+    return ast->GetSwiftType(type);
+  return {};
 }
 
-swift::Type GetSwiftType(const CompilerType &type) {
-  return GetSwiftType(reinterpret_cast<void *>(type.GetOpaqueQualType()));
+swift::CanType GetCanonicalSwiftType(CompilerType type) {
+  auto *ts = type.GetTypeSystem();
+  if (auto *tr = llvm::dyn_cast_or_null<TypeSystemSwiftTypeRef>(ts))
+    return tr->GetSwiftType(type)->getCanonicalType();
+  if (auto *ast = llvm::dyn_cast_or_null<SwiftASTContext>(ts))
+    return ast->GetSwiftType(type)->getCanonicalType();
+  return swift::CanType();
```

Comment:
we should assert we never hit this -- we always should have a valid canonical type at this point if the AST isn't poisoned.

Reply:
No, we can't do that. It can happen when we failed to load a .swiftmodule. Previously this would have resulted in an empty lldb::Type, but because we are deferring type reconstruction, we only find out about it here. But that's not a reason to crash/asssert.